### PR TITLE
Use a Wider Range of Prices

### PR DESCRIPTION
**Pull request by bh2smith**
In order to widen the range on token prices we have for the solver token imbalance we use the median price of all prices for each relevant token for the entire day (Monday) before the token payout (EndTime).

To accompany this change with some data, here is a query showing some stats on the prices we have obtained (showing min_price, max_price, deviation, and range_ratio as max/min)

https://dune.com/queries/888236

We notice some very large gaps in some of the top tokens sorted by range_ratio descending. 

But it appears these were indeed the prices

https://dune.com/queries/888338?EndTime_d83555=2022-06-07+00%3A00%3A00

### CODE REVIEW

File: queries/period_slippage.sql [hunk 1]
```diff
@@ -319,12 +319,16 @@ final_token_balance_sheet as (
...
+    select
+        contract_address,
+        decimals,
+        percentile_cont(0.5) within group (order by median_price) as price
+    from prices.prices_from_dex_data p
+    where hour > '{{EndTime}}'::timestamptz - interval '1 day'
+    and date(hour) = '{{EndTime}}'::timestamptz - interval '1 day'
+    and contract_address in (select token from final_token_balance_sheet)
+    group by contract_address, decimals
+    having sum(sample_size) > 0
```
Comment (josojo) [inline queries/period_slippage.sql:344]: Maybe we can do the following:

- if there is a price feed from prices.usd then we take this price feed and not this volatile price feed from here. this gives us more accuracy, if we have very recent price changes, as we don't average over 1 day
- if prices.usd does not have a price, we take your query. I don't know what to do about the big price ranges. But I guess its anyways mostly outlies and we are same from them due to the precentile_cont command.

I can't come up with a much better method :) sorry.

Comment (bh2smith) [reply queries/period_slippage.sql:344]: Yes we're can do this. I am a bit surprised that a change like this is not affecting our unit tests. I guess our tests don't test the prices but only the token imbalance.

**Comment by bh2smith** [comment]
@josojo - I have updated the query to use the hybrid prices approach. This is the CSV we would have gotten from last week's payout with it.

```
token_type,token_address,receiver,amount
erc20,0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB,0x6C642caFCbd9d8383250bb25F67aE409147f78b2,150900.0
erc20,0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB,0xf7f9C5ecAFd159271fe55A8FEdEDa95d7fcD0DCc,92100.0
erc20,0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB,0x84dbAE2549d67CaF00f65C355de3D6F4df59A32C,41000.0
native,,0xdE1c59Bc25D806aD9DdCbe246c4B5e5505645718,11.635067070979549
native,,0xE58c68679e7AAb8ef83BF37E88b18eB1f6E30e22,11.08113845038937
native,,0x6fa201C3Aff9f1e4897Ed14c7326cF27548d9c35,7.990189916677067
native,,0xe92F359e6F05564849AFa933CE8F62b8007A1d5D,6.65347616185745
native,,0x77ec2A722c2393D3fD64617BBaF1499C713e616b,4.9118455810986354
native,,0x15F4c337122Ec23859EC73Bec00aB38445E45304,3.335076988426351
native,,0x2d15894FaC906386fF7f4Bd07fcEAc43fcF80C73,2.531405318347745
native,,0x833f076D182123cA8DDE2743045Ea02957bD61ef,1.6936513060068359
native,,0xe33062a24149f7801a48B2675Ed5111d3278F0f5,0.33674223075012377
native,,0xa97DEF4fBCBa3b646dd169Bc2eeE40f0f3fE7771,0.2980168390056737
native,,0x109BF9E0287Cc95cc623FBE7380dD841d4bdEb03,0.20070030056888355
native,,0xDe786877a10DBb7EBa25a4DA65aEcf47654F08ab,0.17971803912145667
native,,0xdAE69AffE582d36f330Ee1145995A53Fab670962,0.1654108308648027
native,,0x5B0BFe439ab45A4F002C259b1654ED21c9a42D69,0.14811929193755582
native,,0xda324c2f06d3544E7965767ce9ca536dcb67a660,0.1380498171051335
native,,0x080a8b1E2f3695E179453c5e617b72A381BE44B9,0.12713397491958905
native,,0x340185114F9D2617DC4c16088d0375D09fee9186,0.12276024978485647
native,,0x0d2584DA2F637805071f184bCFa1268eBAe8a24A,0.08174607860512134
native,,0x8ccc61dBA297833dbe5D95FD6360f106b9A7576e,0.009385460612934915
native,,0xf2D21ad3c88170D4AE52bBBeBA80Cb6078D276F4,0.008375422196599536
```


In comparison with the one we actually used:

```
token_type,token_address,receiver,amount
erc20,0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB,0x6C642caFCbd9d8383250bb25F67aE409147f78b2,150900.0
erc20,0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB,0xf7f9C5ecAFd159271fe55A8FEdEDa95d7fcD0DCc,92100.0
erc20,0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB,0x84dbAE2549d67CaF00f65C355de3D6F4df59A32C,41000.0
native,,0xdE1c59Bc25D806aD9DdCbe246c4B5e5505645718,11.650515396216436
native,,0xE58c68679e7AAb8ef83BF37E88b18eB1f6E30e22,8.939991949419674
native,,0x6fa201C3Aff9f1e4897Ed14c7326cF27548d9c35,7.990189916677067
native,,0xe92F359e6F05564849AFa933CE8F62b8007A1d5D,6.65347616185745
native,,0x77ec2A722c2393D3fD64617BBaF1499C713e616b,4.9118455810986354
native,,0x15F4c337122Ec23859EC73Bec00aB38445E45304,3.3359981195478055
native,,0x2d15894FaC906386fF7f4Bd07fcEAc43fcF80C73,2.531405318347745
native,,0x833f076D182123cA8DDE2743045Ea02957bD61ef,1.6936513060068359
native,,0xe33062a24149f7801a48B2675Ed5111d3278F0f5,0.33674223075012377
native,,0xa97DEF4fBCBa3b646dd169Bc2eeE40f0f3fE7771,0.2980168390056737
native,,0x109BF9E0287Cc95cc623FBE7380dD841d4bdEb03,0.20070030056888355
native,,0xDe786877a10DBb7EBa25a4DA65aEcf47654F08ab,0.17971803912145667
native,,0xdAE69AffE582d36f330Ee1145995A53Fab670962,0.1654108308648027
native,,0x5B0BFe439ab45A4F002C259b1654ED21c9a42D69,0.14139151546608436
native,,0xda324c2f06d3544E7965767ce9ca536dcb67a660,0.13804951562186454
native,,0x080a8b1E2f3695E179453c5e617b72A381BE44B9,0.1271165117538005
native,,0x340185114F9D2617DC4c16088d0375D09fee9186,0.12276024978485647
native,,0x0d2584DA2F637805071f184bCFa1268eBAe8a24A,0.08174607860512134
native,,0x8ccc61dBA297833dbe5D95FD6360f106b9A7576e,0.009385460612934915
native,,0xf2D21ad3c88170D4AE52bBBeBA80Cb6078D276F4,0.008375422196599536
```

The most relevant difference being `11.08 - 8.94 = 2.04` for `0xE58c68679e7AAb8ef83BF37E88b18eB1f6E30e22`. We reimbursed `1.126` which suggests there was another incorrect batch for this solver with the previous prices.

I did some analysis on the integrity of using these prices:

- Of the relevant tokens there were 25 still without prices
- the ratio of precise average prices has only 1 extreme outlier (FOX token 0xc770eefad204b5180df6a14ee197d99d808ee52d)
- all other prices are within 25% of each other.
- exact / approx has mean of 0.97 (exact price is 3% below approx) and median of 0.997 and standard deviation of 0.129

Multi Query with all these results is here:
https://dune.com/queries/888236

**Review by josojo — APPROVED** [review]
Great. Thanks for making the comparison with the previous payout.

I think that this solution + slippage alters should be good enough to prevent huge errors. 
But as you show, its still not perfect and let's wait until we concluded a better solution

**Comment by bh2smith** [comment]
As a follow up PR to this, I would like to pull the solver accounting dashboard along with some outlier queries into this repo. Then update the readme with some debugging/investigation procedure. 

Not exactly sure when I will get a chance to do this, but ideally before next payout.

**Pull request merged**